### ISSUE TITLE
Interactive (REPL) support in SparkCLR (for integration with higher-level tools - Jupyter)

### ISSUE THREAD

**Issue by ramangrover** [Microsoft/Mobius#117]
SparkCLR needs to support the interactive mode that enables the end-user to express computation/analysis as a set of C# code statements that get executed in the REPL  (Read, Evaluate, Print, Loop) way. An end-user could then interact with SparkCLR via a prompt. Interactive mode also provides a stepping stone for integration with high-level tools - particularly Jupyter. 

Summarizing design considerations below. 
High-level architecture/design for Jupyter/SparkCLR integration. 

•	SparkCLR’s launch mechanism (sparkclr-submit.cmd) gets a flag (-i, interactive mode). In the interactive mode, SparkCLR (actually CSharpRunner) spawns a RESTful server that has the following REST-based APIs.
1.	Start Session – String Start(SparkConf conf)
2.	End Session – void End(string sessionId)
3.	Execute – ResultSet Execute(String sessionId, String code)

•	Each SparkCLR session has an instance of IREPLEngine, the default implementation being scriptcs. 
             When a session is started, an instance of IREPLEngine is created that supports the Execute call (mirrors the REST-based Execute 
             API in terms of signature). 
             A predefined set of statements is executed that create an instance of SparkContext. This set of statements reflect the standard 
             way of creating SparkContext in SparkCLR. An IREPLEngine instance has an exclusive SparkContext handle. There can be multiple 
             active sessions though, each having its own handle to an IREPLEngine. 

•	An invocation of Execute(string sessionId, string code) receives a C# statement that is passed on to the IREPLEngine instance associated with the session (identified by sessionId) for execution. A statement may produce result that needs to be returned back to the caller. The result is encapsulated inside a ResultSet object which provides an iterator to traverse the individual items.

Integrating with Jupyter
  
Jupyter provides the notion of a “kernel” that is language specific and provides bindings for a given language. A C# kernel is required that allows C# code snippets to be collected from Jupyter’s interface and executed in the backend. ICSharp can be a starting point if  not writing the kernel from scratch. The C# kernel has dependency on SparkCLR and invokes the SparkCLR.cmd launch script in the   
interactive mode. Note that the SparkConfiguration (that determines which cluster user needs to connect with) is received from 
Jupyter and passed on to the C# Kernel and eventually given to SparkCLR as part of arguments to the start script (sparkclr-submit.cmd). 

•	An end-user of Jupyter starts in auto-created session (the session is created by the C# Kernel using the RESTbased interface). By virtue of being in a session, the user has access to SparkContext instance that would be used for executing all statements entered as part of the session. 
•	A statement entered is passed verbatim to the C# Kernel that forwards it using the Execute method call. The Kernel is equipped with handing of the Result set object and presents the result (or exception) to Jupyter as per its protocol. 

To summarize, the tasks can be bucketed into 
SparkCLR
a)	Building the RESTful Server that provides the 3 APIs
b)	Modifying SparkCLR start script to accept interactive mode and launch the server
c)	Building the C# Kernel (possibly on top of ICSharp, source code here)

and (b) can be reused to bridge with Zeppelin in future, if required.

**Comment by toddwan** [comment]
Thanks for the proposal. Below are some initial comments:

* About the RESTful server, I don't think we should spawn it from `CSharpRunner`. Instead, we can implement it as a common Spark app(.exe) powered by SparkCLR. When this sparkclr app starts, it first instantiates a `SparkContext` instance, and then listen to a port waiting for connection from Jupyter, once get connected, executes received c# statements, output results(if any) back to client, and exit when receive a shutdown command.

* If it is reasonable to treat the RESTful server as a common SparkCLR app, `-i` option might not be required. However, it might be convenient for some special requirements like specifying a port if we have a dedicate cmd `sparkclr-rest.cmd` - just a thought.

* `An IREPLEngine instance has an exclusive SparkContext handle. There can be multiple active sessions though, each having its own handle to an IREPLEngine`
If I don't miss anything, do you mean that there might be multiple SparkContext instances in one process?
According to https://issues.apache.org/jira/browse/SPARK-2243, currently spark does not support multiple SparkContext instances in one JVM process. This means that one RESTful server can only support one session - if we don't consider reusing here. We might need to figure out how to manage multiple rest server processes to support multiple sessions in Jupyter side.
Please note that Zeppelin uses one SparkContext instance to run the statements from all users. Zeppelin just guarantees that statements from one user won't be interrupted once start to execute.

**Comment by tawan0109** [comment]
Thanks for the proposal. Below are some initial comments:

* About the RESTful server, I don't think we should make code changes to spawn it from `CSharpRunner`. Instead, we can implement it as a common Spark app(.exe) powered by SparkCLR. When this sparkclr app starts, it first instantiates a `SparkContext` instance, and then listen to a port waiting for connection from Jupyter, once get connected, executes received c# statements, output results(if any) back to client, and exit when receive a shutdown command.

* For a prompt based SparkCLR REPL, we can implement it in the same way, execute the statements from `stdin` and outputs the result to `stdout`.

* If it is reasonable to treat the REPL as a common SparkCLR app, `-i` option might not be required. However, it might be convenient for some special requirements, like specifying a port for RESTful server, if we have a dedicate cmd `sparkclr-rest.cmd` and `sparkclr-shell.cmd`.

* `An IREPLEngine instance has an exclusive SparkContext handle. There can be multiple active sessions though, each having its own handle to an IREPLEngine`
If I don't miss anything, do you mean that there might be multiple SparkContext instances in one process?
According to https://issues.apache.org/jira/browse/SPARK-2243, currently spark does not support multiple SparkContext instances in one JVM process. This means that one RESTful server can only support one session - if we don't consider reusing here. We might need to figure out how to manage multiple rest server processes to support multiple sessions in Jupyter side.
Please note that Zeppelin uses one SparkContext instance to run the statements from all users. Zeppelin just guarantees that statements from one user won't be interrupted once start to execute.

**Comment by danielli90** [comment]
Thank @ramangrover for the proposal. I agreed with @tawan0109 that SparkCLRShell can be implemented as a regular SparkCLR app, leveraging scriptcs, without RESTful server involved, similar to how PySpark shell is implemented.

Agreed that C# Kernel is needed for Jupyter integration, which can be built on top of SparkCLRShell. For example, Jupyter IPython kernel submits a PysparkShell job in Spark - http://blog.insightdatalabs.com/jupyter-on-apache-spark-step-by-step/.

**Comment by skaarthik** [comment]
@ramangrover - thanks for the great write-up.

Looking forward to your response to the comments from @tawan0109. 
It would be great if you can add comments on alternative architectures you considered. For example, would it help to have a standalone REST server that leverage SparkCLR API? - like an ASP.NET Web API service that does the exact same thing you explained under (a). I think that would be similar to the app suggestion from @tawan0109 but with the benefits of a server/service.

For (c), I am not quite clear on what is required from our side on top of the features available in ICSharp (or scriptCS). Please add some details.

**Comment by skaarthik** [comment]
@tawan and @kai-zeng completed the work and it is already merged
